consensus/src/da_replica.rs

```diff
@@ -0,0 +1,526 @@
+            match &parent_view.view_inner {
+                ViewInner::Leaf { leaf } => {
+                    if let Some(leaf) = consensus.saved_leaves.get(leaf) {
+                        leaf
+                    } else {
+                        warn!("Failed to find high QC parent.");
+                        return None;
+                    }
+                }
+                ViewInner::Failed => {
+                    warn!("Parent of high QC points to a failed QC");
+                    return None;
+                }
+            }
+        } else {
+            warn!("Couldn't find high QC parent in state map.");
+            return None;
+        };
+        Some(parent_leaf.clone())
+    }
+
+    /// Replica task for sequencing consensus that spins until a vote can be made or timeout is
+    /// hit.
+    ///
+    /// Returns the new leaf if it's valid.
+    #[instrument(skip_all, fields(id = self.id, view = *self.cur_view), name = "Sequencing Replica Task", level = "error")]
+    #[allow(clippy::type_complexity)]
+    async fn find_valid_msg<'a>(
+        &self,
+        view_leader_key: TYPES::SignatureKey,
+        consensus: RwLockUpgradableReadGuard<'a, Consensus<TYPES, DALeaf<TYPES>>>,
+    ) -> (
+        RwLockUpgradableReadGuard<'a, Consensus<TYPES, DALeaf<TYPES>>>,
+        Option<DALeaf<TYPES>>,
+    ) {
+        let lock = self.proposal_collection_chan.lock().await;
+        let mut invalid_qc = false;
+        let leaf = loop {
+            let msg = lock.recv().await;
+            info!("recv-ed message {:?}", msg.clone());
+            if let Ok(msg) = msg {
+                // stale/newer view messages should never reach this specific task's receive channel
+                if Into::<ConsensusMessage<_, _, _>>::into(msg.clone()).view_number()
+                    != self.cur_view
+                {
+                    continue;
+                }
+                match msg {
+                    ProcessedConsensusMessage::Proposal(p, sender) => {
+                        if view_leader_key != sender {
+                            continue;
+                        }
+
+                        let mut valid_leaf = None;
+                        let vote_token = self.api.make_vote_token(self.cur_view);
+                        match vote_token {
+                            Err(e) => {
+                                error!(
+                                    "Failed to generate vote token for {:?} {:?}",
+                                    self.cur_view, e
+                                );
+                            }
+                            Ok(None) => {
+                                info!("We were not chosen for committee on {:?}", self.cur_view);
+                            }
+                            Ok(Some(vote_token)) => {
+                                info!("We were chosen for committee on {:?}", self.cur_view);
```

Comment:
nit: maybe the logging should specify the committee type, since here the committe is bascially anyone with stake, but in DAMember it's really checking membership on the DA comittee